---
title: "nf-core/taxprofiler: highly parallelised and flexible pipeline for metagenomic taxonomic classification and profiling"
author: 
    - name: Sofia Stamouli
      affiliations:
      - name: Karolinkska Institutet
    - name: Moritz Beber
      affiliations:
      - name: UnSeen Bio
    - name: "..."
    - name: James A. Fellows Yates
      orcid: 0000-0001-5585-6277
      email: james_fellows_yates@eva.mpg.de
      affiliations:
      - name: Max Planck Institute for Evolutionary Anthropology
        address: Deutsche Pl. 6 
        city: Leipzig
        state: Saxony
        postal-code: 04103
bibliography: refs.bib
---

## Abstract

Metagenomic classification tackles the problem of characterising the taxonomic source of all DNA sequencing reads in a sample. A common approach to address the differences and biases between the many different taxonomic classification tools is to run sequencing libraries through multiple classification tools and databases. This, however, is a very time intensive task when performing manually - particularly when combined with the appropriate preprocessing of sequencing reads before the classification. 

Here we present nf-core/taxprofiler, a highly parallelised taxonomic classification and processing pipeline that allows for automated and simultaneous classification and/or profiling of of both short- and long-read metagenomic sequencing libraries against a large number of taxonomic classifiers and profilers as well as databases in one pipeline run. Implemented in Nextflow and nf-core, the pipeline benefits from high levels of scalability and portability, allowing for large and small projects on a wide range of computing infrastructure, as well as best-practise software development and community support to ensure longevity and adaptability of the pipeline as the field of metagenomics develops. <...>

## Introduction


There are a range of strong benefits of metagenomic approaches to taxonomic classification of DNA samples compared to targeted approaches [CITE]. While targeted approaches such as 16S or other marker gene amplicon sequencing have in the past been useful due to low cost and large and diverse reference databases, metagenomic analyses has been shown to produce similar resolution during taxonomic classification [@Hillmann2018-ud] (with increasing lower costs of shotgun sequencing), but the added benefit of having greater reusability potential of the data, via whole genome reconstruction and also functional classification of metagenomics.

<!-- TODO: define our classification/profiling terimnology? -->

Taxonomic profiling consists of identifying the original 'taxonomic source' of a given DNA sequence [CITE]. In metagenomics this typically consists of comparing millions of DNA sequences against hundreds or thousands of reference genomes either via alignment or 'k-mer matching' [CITE], and the most close match being considered the most likely original 'source' organism of that sequence. Some profilers additionally will also try to infer actual (cellular) species abundance of the organism in the original sample, based on the sequence abundance [CITE]. 

Due to the scale of taxonomic profiling is an unresolved 'problem' in bioinformatics. Having to identify the original source of many sequences, from many samples, against many reference genomes but in an _efficient_ manner is understandably a difficult problem. Therefore a plethora of tools have been developed to address this challenge, all with their own biases and specific contexts that they have been developed in [@Sczyrba2017-ma;@Meyer2022-dg]. Thus, no established 'gold standard' method currently exists.

<!-- ONE SOLUTION IS TO RUN MANY -->
<!-- DATABASES? -->

<!-- SCALABILITY DUE TO LARGE SCALE -->

## Implementation

![Visual overview of the nf-core/taxprofiler workflow. nf-core/taxprofiler can take in FASTQ (short or long reads) or FASTA files (long reads), that will optionally go through sequencing quality control, read preprocessing, complexity filtering, host removal, and run merging before going into taxonomic classification and/or profiling with a user-selected range of tools and databases. Output from all classifiers and profilers are additionally standardised into a common taxon table format and some with visualisation of profiles.](taxprofiler_tube.png){#fig-workflow-diagram fig-alt="nf-core/taxprofiler workflow diagram in a metro map style"}

There are many types of metagenomic profile techniques from alignment, to k-mer, and marker-gene based methods [@Quince2017-bl]

## Use Cases

## Discussion

## Data Availability

## Supplementary Data

## Acknowledgments

## Funding

## References


Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto see <https://quarto.org>.
