---
title: "nf-core/taxprofiler: highly parallelised and flexible pipeline for metagenomic taxonomic classification and profiling"
## TODO: author roels: https://quarto.org/docs/authoring/front-matter.html#roles
author: 
    - name: Sofia Stamouli
      given: Sofia
      family: Stamouli
      orcid: 0009-0006-0893-3771
      affiliations:
      - ref: ki
      - ref: kuh
    - name: Moritz Beber
      given: Moritz
      family: Beber
      orcid: 0000-0003-2406-1978
      affiliations:
      - ref: usb
    - name: Lauri Mesilaakso
      given: Lauri
      family: Mesilaakso
      orcid:
      affiliations:
      - ref: rog
    - name: Tanja Normark
      given: Tanja 
      family: Normark
      orcid:
      affiliations:
      - ref: ki 
      - ref: kuh
    - name: Jianhong Ou
      given: Jianhong
      family: Ou
      orcid:
      affiliations:
      - ref: du
    - name: Thomas A. Christensen II
      given: Thomas A.
      family: Christensen II
      orcid: 0000-0003-1219-9320
      affiliations:
      - ref: ksu
    - name: Maxime Borry
      given: Maxime
      family: Borry
      orcid: 0000-0001-9140-7559
      affiliations:
      - ref: eva
    - name: Mahwash Jamy
      given: Mahwash
      family: Jamy
      orcid:
      affiliations:
      - ref: ki
      - ref: kuh
    - name: Rafal Stepien
      given: Rafal
      family: Stepien
      orcid:
      affiliations:
      - ref: uv
    - name: nf-core community
      url: https://nf-co.re
      affiliations:
        - ref: xx
    - name: James A. Fellows Yates
      given: James A.
      family: Fellows Yates
      orcid: 0000-0001-5585-6277
      email: james_fellows_yates@eva.mpg.de
      corresponding: true
      affiliations:
        - ref: eva
        - ref: hki
affiliations:
  - id: ki
    name: Department of Microbiology, Tumor and Cell Biology, Clinical Genomics, Karolinkska Institutet
    address: Solnavägen 1
    city: Solna
    state: Stockholm 
    postal-code: 17177
    country: Sweden
  - id: kuh
    name: Department of Clinical Microbiology, Karolinska University Hospital
    address: Anna Steckséns gata 41
    city: Solna
    state: Stockholm
    postal-code: 17177
    country: Sweden 
  - id: usb
    name: Unseen Bio ApS
    address: Fruebjergvej 3
    city: Copenhagen
    state: Capital Region
    postal-code: 2100
    country: Denmark
  - id: eva
    name: Department of Archaeogenetics, Max Planck Institute for Evolutionary Anthropology
    address: Deutsche Pl. 6 
    city: Leipzig
    state: Saxony
    postal-code: 04103
    country: Germany
  - id: hki
    name: Department of Paleobiotechnology, Leibniz Institute for Natural Product Research and Infection Biology Hans Knöll Institute
    address: Adolf-Reichwein-Straße
    city: Jena
    state: Thuringen
    postal-code: 07745
    country: Germany
  - id: rog
    name: Department of Clinical Microbiology, Region Östergötland
    address: "S:t Larsgatan	49 B"
    city: Linköping
    state: Östergötland
    postal-code: 58191
    country: Sweden
  - id: du
    name: Regeneration Center, Duke University
    address: Nanaline Duke Bldg 464
    city: Durham
    state: NC
    postal-code: 27710
    country: United States
  - id: ksu
    name: Veterinary Diagnostic Laboratory, Kansas State University College of Veterinary Medicine 
    address: 1800 Denison Ave
    city: Manhattan
    state: KS
    postal-code: 66506
    country: United States
  - id: uv
    name: Center for Public Health Genomics, University of Virginia
    address: University Avenue 1827
    city: Charlottesville
    state: VA
    postal-code: 22903
    country: United States
  - id: xx
bibliography: refs.bib
license: "CC BY"
---

## Abstract

Metagenomic classification tackles the problem of characterising the taxonomic source of all DNA sequencing reads in a sample. A common approach to address the differences and biases between the many different taxonomic classification tools is to run sequencing libraries through multiple classification tools and databases. This, however, is a very time intensive task when performing manually - particularly when combined with the appropriate preprocessing of sequencing reads before the classification. 

Here we present nf-core/taxprofiler, a highly parallelised taxonomic classification and processing pipeline that allows for automated and simultaneous classification and/or profiling of of both short- and long-read metagenomic sequencing libraries against a large number of taxonomic classifiers and profilers as well as databases in one pipeline run. Implemented in Nextflow and nf-core, the pipeline benefits from high levels of scalability and portability, allowing for large and small projects on a wide range of computing infrastructure, as well as best-practise software development and community support to ensure longevity and adaptability of the pipeline as the field of metagenomics develops.

## Introduction

There are a range of strong benefits of metagenomic approaches to taxonomic classification of DNA samples compared to targeted approaches [CITE]. While targeted approaches such as 16S or other marker gene amplicon sequencing have in the past been useful due to low cost and large and diverse reference databases, metagenomic analyses has been shown to produce similar resolution during taxonomic classification [@Hillmann2018-ud] (with increasing lower costs of shotgun sequencing), but the added benefit of having greater reusability potential of the data, via whole genome reconstruction and also functional classification of metagenomics.

<!-- TODO: define our classification/profiling terimnology? -->

Taxonomic profiling consists of identifying the original 'taxonomic source' of a given DNA sequence [CITE]. In metagenomics this typically consists of comparing millions of DNA sequences against hundreds or thousands of reference genomes either via alignment or 'k-mer matching' [CITE], and the most close match being considered the most likely original 'source' organism of that sequence. Some profilers additionally will also try to infer actual (cellular) species abundance of the organism in the original sample, based on the sequence abundance [CITE]. 

Due to the scale of the problem, taxonomic profiling remains an 'unresolved problem' in bioinformatics. Having to identify the original source of many sequences, against many reference genomes but in an _efficient_ manner is understandably a difficult problem. Therefore a plethora of tools have been developed to address this challenge, all with their own biases and specific contexts that they have been developed in [@Sczyrba2017-ma;@Meyer2022-dg]. Additionally, each tool often produces tool-specific output formats making it difficult to efficiently cross compare results. Thus, no established 'gold standard' method currently exists.

One solution to addressing the range of different tools is to run all of them in parallel, and cross compare the results. This can both be useful for benchmarking studies [e.g. @Sczyrba2017-ma;@Meyer2022-dg], but also to build consensus profiles whereby confidence of a particular taxonomic identification can be increased when it is detected by multiple tools [CITE].

A second challenge in taxonomic profiling is a question of databases. As with tools, there is no one set 'gold standard' database for taxonomic profiling. Different questions and contexts may require different databases, such as when a researcher wants to search for both bacteria_and viruses species in samples, and as an extension of this, classifiers may need different settings for each database. Furthermore, as genomic sequencing becomes cheaper and more efficient, the number of publicly available reference genomes are rapidly increasing [CITE], making the size of databases taxonomic classifiers also much a larger. 

Finally, with the sinking of sequencing costs, this opens up the possibility for larger and larger metagenome sequencing, meaning the sample sizes of studies are increasing further highlighting the need for methods for efficiently profiling many samples against many tools and databases. Manually setting up bioinformatic jobs for classification tasks for each database and settings against different tools on traditional academic computing infrastructure (e.g. high performance computing clusters or 'HPC' clusters) can be very tedious. Additionally, particularly for very large sample sets, there is increasing use of cloud platforms that have greater scalability than traditional HPCs. Being able to reliably and reproducibly execute taxonomic classification tasks across infrastructure with minimal intervention would therefore be a boon for the metagenomics field.

Here we present nf-core/taxprofiler a pipeline designed to allow users to efficiently and simultaneously taxonomically classify and profile short- and long-read sequencing data against multiple tools and databases in a single pipeline run. nf-core/taxprofiler utilises Nextflow [@Di_Tommaso2017-xu] to ensure efficiency, portability, and scalability, and has been developed within the nf-core initiative of Nextflow pipelines [@Ewels2020-vi] to ensure high quality coding practises and user accessibility, including detailed documentation and a GUI execution interface.

## Implementation

nf-core/taxprofiler aims to facilitate three main steps of a typical shotgun metagenomic workflow. Taking in short- (e.g. Illumina) or long-read (e.g. Nanopore) FASTQ or FASTA files, it can perform a range of appropriate preprocessing steps of reads, then perform taxonomic classification profiling against a range of different tools depending on the wish of the user, and finally perform post-classification aggregation and standardisation of the resulting profiles with the possibility of visualisation of outputs [FIGURE 1]. All relevant preprocessing statistics are displayed in an interactive and dynamic MultiQC report [CITE].

![Visual overview of the nf-core/taxprofiler workflow. nf-core/taxprofiler can take in FASTQ (short or long reads) or FASTA files (long reads), that will optionally go through sequencing quality control, read preprocessing, complexity filtering, host removal, and run merging before going into taxonomic classification and/or profiling with a user-selected range of tools and databases. Output from all classifiers and profilers are additionally standardised into a common taxon table format and some with visualisation of profiles.](taxprofiler_tube.png){#fig-workflow-diagram fig-alt="nf-core/taxprofiler workflow diagram in a metro map style"}

### Input and Execution

<!-- TODO comapre cli and gui; laptop vs HPC vs cloud etc. -->
<!-- TODO input sheet and database sheet

<!-- TODO -->

### Preprocessing

Preprocessing steps in nf-core/taxprofiler are aimed at removing laboratory and sequencing artefacts that may influence taxonomically profiling, either for computing resource consumption and/or false-positive or false-negative classification reasons. First sequencing quality control with FastQC [CITE] or Falco [CITE] is carried out. Falco was included in for reduced memory requirements in particular for long reads sequencing. Library adapter sequences can reduce accuracy during alignment and are sometimes contaminants inside reference genomes [CITE CARP BLOG], and paired-end merging may provide longer sequences that will allow more specific classification when paired-end alignment is not supported by a given classifier. For these tasks nf-core/taxprofiler can apply either fastp [CITE] or AdapterRemoval2 [CITE] for short-reads, and Porechop [CITE] for Oxford Nanopore long-read data. For both short- and long-reads FastQC or Falco is run again to allow assessment on the performance of the adapter removal and/or pair-merging step.

Low complexity sequences, e.g. sequences containing long stretches mono- or di-nucleotide repeats provide little specific genetic information that contribute to taxonomic identification, as they can align to many different reference genomes. Including such reads during profiling can increase run-time and memory usage for little gain, as during lowest-common-ancestor (LCA) classification steps they will be assigned to high-level taxonomic levels such as Kingdom. nf-core/taxprofiler offers removal of these reads through complexity filtering algorithms as provided by fastp, BBDuk [CITE], or PRINSEQ++ [CITE]. Long read sequences often do not have such reads as lengths are sufficeint enough to capture greater sequence diversity - but it is sometimes desired to only classify reads longer than a certain length - as these provide sufficient taxonomic information. Therefore nf-core/taxprofiler can remove contigs shorter than a user-defined length using Filtlong.

Another common preprocessing step in metagenomic studies is to remove possible host DNA. This can help speed up run-time, particularly for microbiome studies, where detection of only microbes are of interest. Furthermore, host-contamination of reference genomes on public databases is common [CITE], therefore removal of such sequences can also decrease the risk of false positive taxonomic identifications.  If users wish to remove multiple hosts or other sequences, they can simply combine all reference genomes into a single FASTA reference file. Short-read host removal can be carried out with Bowtie2 [CITE] and minimap2 for long-reads, both in combination with SAMtools [CITE].

Finally, nf-core/taxprofiler can optionally perform run merging where libraries have been sequenced over multiple lanes to generate one profile per sample or library.

Across all steps, relevant statistics and log files are generated, and the processed reads of each step in FASTQ format can be optionally saved for downstream re-use.

### Profiling

There are many types of metagenomic profile techniques from profiling against whole-genome references with alignment or k-mer based approaches [CITE], or methods involving alignment to species-specific marker-gene families [@Quince2017-bl,OTHERS?]. nf-core/taxprofiler aims to support and include all established classification or profiling tools as requested by the community. Which tools are used in a pipeline run is up to the user, with a tool being executed both a corresponding database and `--run_<tool>` flag is provided. Specific classification/profiling settings for each tool and database are specified in the database CSV input sheet, however for some tools pipeline level command-line flags are available. 

As of version 1.1.0, the following classifiers and profilers are available: Kraken2 [CITE], Bracken [CITE], KrakenUniq [CITE], Centrifuge [CITE], MALT [CITE], DIAMOND [CITE], Kaiju [CITE], MetaPhlAn3 [CITE], MetaPhlAn4 [CITE] mOTUs [CITE], ganon [CITE], KMCP [CITE]. Table <!-- TODO --> summarises the category and reference database type for each tool.

<!-- TODO CHECK!!!!! AND add fig id AND maybe add classifier vs profiler? -->
| Method          | Reference Type | Tool       |
|-----------------|----------------|------------|
| k-mer based     | whole-genome   | Kraken2    |
| k-mer based     | whole-genome   | Bracken    |
| k-mer based     | whole-genome   | KrakenUniq |
| k-mer based     | whole-genome   | ganon      |
| ???             | ???            | KMCP       |
| alignment based | whole-genome   | Centrifuge |
| alignment based | whole-genome   | MALT       |
| alignment based | whole-genome   | DIAMOND    |
| alignment based | whole-genome   | Kaiju      |
| alignment based | marker-gene    | MetaPhlAn3 |
| alignment based | marker-gene    | MetaPhlAn4 |
| alignment based | marker-gene    | mOTUS      |

nf-core/taxprofiler by default produces the per-sample main taxonomic classification profile from a tool or a tool's report generation tool, normally in the form of counts per reference sequencing, with additional statistics about the hits of a particular organism (estimated abundance, taxonomic level etc.). Users can also optionally request output of per-read classification output, and output such as classified and unclassified reads in FASTQ format, where supported.

### Post-profiling

Metagenomic studies often must compare the profiles between many samples, and the results of multiple profiles are normally stored in 'taxon tables', i.e, counts per reference (rows), for each sample (columns). When available nf-core/taxprofiler can optionally produce the 'native' taxon table of each classification tool when multiple samples are run. 

When researchers wish to compare between taxonomic classifiers or profilers, they must often deal with the heterogenous output formats that are produced by each tool, typically in the form of custom parser and merging scripts for each tool. To facilitate more user-friendly cross-comparison between tools, nf-core/taxprofiler utilises the TAXPASTA tool [@Beber???] to generate standardarised profiles and generate multi-sample tables.

Summary statistics for the entire pipeline are visualised are displayed in a customisable MultiQC report [CITE], when supported, to support user quality control of data and pipeline runs. Krona plots [CITE] can also optionally be generated for supported tools to help provide further visualisation of taxonomic profiles.

## Use Cases  and Benchmarking

We ran nf-core/taxprofiler on a dataset of <!-- TODO -->  from <!-- TODO --> , on a <!-- TODO --> .

Run metric can be seen in Figure <!-- TODO --> .

## Discussion

Comparison with...

We decided not to include support for metataxonomic tools (amplicon sequencing data) as there already exists well established best-practise pipelines for this [QIIME], however conceptually tools such as XYZ that identify 16S sequences _from_ metagenomic

Future development will see <...additional profiling tools, database construction, additional validation steps, PacBio support>

## Conclusion

## Data Availability

All data used in this publication 

## Code Availability

nf-core/taxprofiler source code is available on GitHub at https://github.com/nf-core/taxprofiler, and each release is archived on Zenodo (<!-- DOI -->)

The version of the pipeline described in this paper is version (<!-- VERSION -->) (Zenodo archive DOI: <!-- DOI -->)

## Supplementary Data

## Acknowledgments

We thank Prof. Christina Warinner and the Microbiome Sciences group MPI-EVA for original discussions that lead to the pipeline. We are also grateful for the nf-core community for ongoing support in the development in the pipeline.

## Funding

S.S. was supported by Rapid establishment of comprehensive laboratory pandemic preparedness – RAPID-SEQ
This material is based upon work supported by the U.S. Department of Agriculture, Agricultural Research Service, under agreement No. 58-3022-0-001 (T.A.C II).
M.B. and J.A.F.Y were supported by the Max Planck Society. J.A.F.Y was supported by the Werner Siemens-Stiftung ("Paleobiotechnology", Awarded to Prof. Pierre Stallforth and Prof. Christina Warinner).


## References