<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Sofia Stamouli">
<meta name="author" content="Moritz Beber">
<meta name="author" content="Lauri Mesilaakso">
<meta name="author" content="Tanja Normark">
<meta name="author" content="Thomas A. Christensen II">
<meta name="author" content="Lili Andersson-Li">
<meta name="author" content="Maxime Borry">
<meta name="author" content="Mahwash Jamy">
<meta name="author" content="Rafal Stepien">
<meta name="author" content="nf-core community">
<meta name="author" content="James A. Fellows Yates">

<title>taxprofiler-manuscript - nf-core/taxprofiler: highly parallelised and flexible pipeline for metagenomic taxonomic classification and profiling</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract">Abstract</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#implementation" id="toc-implementation" class="nav-link" data-scroll-target="#implementation">Implementation</a>
  <ul class="collapse">
  <li><a href="#input-and-execution" id="toc-input-and-execution" class="nav-link" data-scroll-target="#input-and-execution">Input and Execution</a></li>
  <li><a href="#preprocessing" id="toc-preprocessing" class="nav-link" data-scroll-target="#preprocessing">Preprocessing</a></li>
  <li><a href="#profiling" id="toc-profiling" class="nav-link" data-scroll-target="#profiling">Profiling</a></li>
  <li><a href="#post-profiling" id="toc-post-profiling" class="nav-link" data-scroll-target="#post-profiling">Post-profiling</a></li>
  <li><a href="#output" id="toc-output" class="nav-link" data-scroll-target="#output">Output</a></li>
  </ul></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion">Discussion</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  <li><a href="#data-availability" id="toc-data-availability" class="nav-link" data-scroll-target="#data-availability">Data Availability</a></li>
  <li><a href="#code-availability" id="toc-code-availability" class="nav-link" data-scroll-target="#code-availability">Code Availability</a></li>
  <li><a href="#supplementary-data" id="toc-supplementary-data" class="nav-link" data-scroll-target="#supplementary-data">Supplementary Data</a></li>
  <li><a href="#acknowledgments" id="toc-acknowledgments" class="nav-link" data-scroll-target="#acknowledgments">Acknowledgments</a></li>
  <li><a href="#funding" id="toc-funding" class="nav-link" data-scroll-target="#funding">Funding</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">nf-core/taxprofiler: highly parallelised and flexible pipeline for metagenomic taxonomic classification and profiling</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Sofia Stamouli <a href="https://orcid.org/0009-0006-0893-3771" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Microbiology, Tumor and Cell Biology, Karolinkska Institutet
          </p>
        <p class="affiliation">
            Department of Clinical Microbiology, Karolinska University Hospital
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Moritz Beber <a href="https://orcid.org/0000-0003-2406-1978" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Unseen Bio ApS
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Lauri Mesilaakso </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Clinical Microbiology, Region Östergötland
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Tanja Normark </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Microbiology, Tumor and Cell Biology, Karolinkska Institutet
          </p>
        <p class="affiliation">
            Department of Clinical Microbiology, Karolinska University Hospital
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Thomas A. Christensen II <a href="https://orcid.org/0000-0003-1219-9320" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Veterinary Diagnostic Laboratory, Kansas State University College of Veterinary Medicine
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Lili Andersson-Li </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Microbiology, Tumor and Cell Biology, Karolinkska Institutet
          </p>
        <p class="affiliation">
            Department of Clinical Microbiology, Karolinska University Hospital
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Maxime Borry <a href="https://orcid.org/0000-0001-9140-7559" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Archaeogenetics, Max Planck Institute for Evolutionary Anthropology
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Mahwash Jamy </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Microbiology, Tumor and Cell Biology, Karolinkska Institutet
          </p>
        <p class="affiliation">
            Department of Clinical Microbiology, Karolinska University Hospital
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Rafal Stepien </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Center for Public Health Genomics, University of Virginia
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author"><a href="https://nf-co.re">nf-core community</a> </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">James A. Fellows Yates <a href="https://orcid.org/0000-0001-5585-6277" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Archaeogenetics, Max Planck Institute for Evolutionary Anthropology
          </p>
        <p class="affiliation">
            Department of Paleobiotechnology, Leibniz Institute for Natural Product Research and Infection Biology Hans Knöll Institute
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  

</header>

<section id="abstract" class="level2">
<h2 class="anchored" data-anchor-id="abstract">Abstract</h2>
<p>Metagenomic classification tackles the problem of characterising the taxonomic source of all DNA sequencing reads in a sample. A common approach to address the differences and biases between the many different taxonomic classification tools is to run sequencing libraries through multiple classification tools and databases. This, however, is a very time-consuming task when performing manually - particularly when combined with the appropriate preprocessing of sequencing reads prior to classification.</p>
<p>Here we present nf-core/taxprofiler, a highly parallelised taxonomic classification and processing pipeline that allows for automated and simultaneous classification and/or profiling of both short- and long-read metagenomic sequencing libraries against a large number of taxonomic classifiers and profilers as well as databases within a single pipeline run. Implemented in Nextflow and as part of the nf-core initiative, the pipeline benefits from high levels of scalability and portability, allowing for large and small projects on a wide range of computing infrastructure, as well as best-practise software development and community support to ensure longevity and adaptability of the pipeline, keeping up with the evolving field of metagenomics.</p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Shotgun metagenomics offers strong benefits to the taxonomic classification of DNA samples over targeted approaches <span class="citation" data-cites="Eloe-Fadrosh2016-gz Breitwieser2019-ys">(<a href="#ref-Eloe-Fadrosh2016-gz" role="doc-biblioref">Eloe-Fadrosh et al. 2016</a>; <a href="#ref-Breitwieser2019-ys" role="doc-biblioref">Florian P. Breitwieser, Lu, and Salzberg 2019</a>)</span>. While targeted approaches such as 16S or other marker gene amplicon sequencing have in the past been useful due to low cost and large and diverse reference databases <span class="citation" data-cites="Yilmaz2014-js Lynch2015-mz">(<a href="#ref-Yilmaz2014-js" role="doc-biblioref">Yilmaz et al. 2014</a>; <a href="#ref-Lynch2015-mz" role="doc-biblioref">Lynch and Neufeld 2015</a>)</span>, metagenomic analyses has been shown to produce similar resolution during taxonomic classification <span class="citation" data-cites="Hillmann2018-ud">(<a href="#ref-Hillmann2018-ud" role="doc-biblioref">Hillmann et al. 2018</a>)</span> (with increasing lower costs of shotgun sequencing), but the added benefit of having greater reusability potential of the data, via whole genome reconstruction and also functional classification of metagenomics <span class="citation" data-cites="Sharpton2014-wy Quince2017-bl">(<a href="#ref-Sharpton2014-wy" role="doc-biblioref">Sharpton 2014</a>; <a href="#ref-Quince2017-bl" role="doc-biblioref">Quince et al. 2017</a>)</span>.</p>
<p>Taxonomic profiling consists of identifying the original ‘taxonomic source’ of a given DNA sequence <span class="citation" data-cites="Ye2019-fl Govender2022-td">(<a href="#ref-Ye2019-fl" role="doc-biblioref">Ye et al. 2019</a>; <a href="#ref-Govender2022-td" role="doc-biblioref">Govender and Eyre 2022</a>)</span>. In metagenomics this typically consists of comparing millions of DNA sequences against hundreds or thousands of reference genomes either via alignment or ‘k-mer matching’ <span class="citation" data-cites="Sharpton2014-wy Sun2021-tq">(<a href="#ref-Sharpton2014-wy" role="doc-biblioref">Sharpton 2014</a>; <a href="#ref-Sun2021-tq" role="doc-biblioref">Sun et al. 2021</a>)</span>, and the most close match being considered the most likely original ‘source’ organism of that sequence. Some profilers additionally will also try to infer actual (cellular) species abundance of the organism in the original sample, based on the sequence abundance <span class="citation" data-cites="Nayfach2016-ej">(<a href="#ref-Nayfach2016-ej" role="doc-biblioref">Nayfach and Pollard 2016</a>)</span>. <!--
I see that this paragraph has already been adressed.
--></p>
<p>Due to the scale of the problem, taxonomic profiling remains an ‘unresolved problem’ in bioinformatics. Having to identify the original source of many sequences, against many reference genomes but in an <em>efficient</em> manner is understandably a difficult problem. Therefore a plethora of tools have been developed to address this challenge, all with their own biases and specific contexts that they have been developed in <span class="citation" data-cites="Sczyrba2017-ma Meyer2022-dg">(<a href="#ref-Sczyrba2017-ma" role="doc-biblioref">Sczyrba et al. 2017</a>; <a href="#ref-Meyer2022-dg" role="doc-biblioref">Meyer et al. 2022</a>)</span>. Additionally, each tool often produces tool-specific output formats making it difficult to efficiently cross compare results. Thus, no established ‘gold standard’ method currently exists.</p>
<p>One solution to addressing the range of different tools is to run all of them in parallel, and cross compare the results. This can both be useful for benchmarking studies <span class="citation" data-cites="Sczyrba2017-ma Meyer2022-dg">(e.g. <a href="#ref-Sczyrba2017-ma" role="doc-biblioref">Sczyrba et al. 2017</a>; <a href="#ref-Meyer2022-dg" role="doc-biblioref">Meyer et al. 2022</a>)</span>, but also to build consensus profiles whereby confidence of a particular taxonomic identification can be increased when it is detected by multiple tools <span class="citation" data-cites="McIntyre2017-td Ye2019-fl">(<a href="#ref-McIntyre2017-td" role="doc-biblioref">McIntyre et al. 2017</a>; <a href="#ref-Ye2019-fl" role="doc-biblioref">Ye et al. 2019</a>)</span>.</p>
<p>A second challenge in taxonomic profiling is a question of databases. As with tools, there is no one set ‘gold standard’ database for taxonomic profiling. Different questions and contexts may require different databases, such as when a researcher wants to search for both bacteria_and viruses species in samples, and as an extension of this, classifiers may need different settings for each database. Furthermore, as genomic sequencing becomes cheaper and more efficient, the number of publicly available reference genomes are rapidly increasing <span class="citation" data-cites="Nasko2018-cl">(<a href="#ref-Nasko2018-cl" role="doc-biblioref">Nasko et al. 2018</a>)</span>, making the size of databases taxonomic classifiers also much a larger and often out pacing the computational capacity available to researchers. In fact, while this was one of the main motivations behind classifiers such as Kraken2 <span class="citation" data-cites="Wood2019-mf">(<a href="#ref-Wood2019-mf" role="doc-biblioref">Wood, Lu, and Langmead 2019</a>)</span>, these algorithmic techniques are already becoming insufficient <span class="citation" data-cites="Wright2023-yr">(<a href="#ref-Wright2023-yr" role="doc-biblioref">Wright, Comeau, and Langille 2023</a>)</span>.</p>
<p>Finally, with the sinking of sequencing costs, this opens up the possibility for larger and larger metagenome sequencing, meaning the sample sizes of studies are increasing further highlighting the need for methods for efficiently profiling many samples against many tools and databases - as exemplified by the doubling of the number of metagenomes on the European Bioinformatic Institute’s MGnify database in two years <span class="citation" data-cites="Mitchell2019-kt">(<a href="#ref-Mitchell2019-kt" role="doc-biblioref">Mitchell et al. 2019</a>, `)</span>. Manually setting up bioinformatic jobs for classification tasks for each database and settings against different tools on traditional academic computing infrastructure (e.g.&nbsp;high performance computing clusters or ‘HPC’ clusters) can be very tedious. Additionally, particularly for very large sample sets, there is increasing use of cloud platforms that have greater scalability than traditional HPCs. Being able to reliably and reproducibly execute taxonomic classification tasks across infrastructure with minimal intervention would therefore be a boon for the metagenomics field.</p>
<p>Here we present nf-core/taxprofiler, a pipeline designed to allow users to efficiently and simultaneously taxonomically classify and profile short- and long-read sequencing data against multiple tools and databases in a single pipeline run. nf-core/taxprofiler utilises Nextflow <span class="citation" data-cites="Di_Tommaso2017-xu">(<a href="#ref-Di_Tommaso2017-xu" role="doc-biblioref">Di Tommaso et al. 2017</a>)</span> to ensure efficiency, portability, and scalability, and has been developed within the nf-core initiative of Nextflow pipelines <span class="citation" data-cites="Ewels2020-vi">(<a href="#ref-Ewels2020-vi" role="doc-biblioref">Ewels et al. 2020</a>)</span> to ensure high quality coding practises and user accessibility, including detailed documentation and a graphical-user-interface (GUI) execution interface.</p>
</section>
<section id="implementation" class="level2">
<h2 class="anchored" data-anchor-id="implementation">Implementation</h2>
<p>nf-core/taxprofiler aims to facilitate three main steps of a typical shotgun metagenomic workflow. Taking in short- (e.g.&nbsp;Illumina) or long-read (e.g.&nbsp;Nanopore) FASTQ or FASTA files, it can perform a range of appropriate preprocessing steps of reads, then perform taxonomic classification and profiling against a range of different tools depending on the wish of the user, and finally perform post-classification aggregation and standardisation of the resulting profiles with the possibility of visualisation of outputs [FIGURE 1]. All relevant preprocessing statistics are displayed in an interactive and dynamic MultiQC report <span class="citation" data-cites="Ewels2020-vi">(<a href="#ref-Ewels2020-vi" role="doc-biblioref">Ewels et al. 2020</a>)</span>.</p>
<div id="fig-workflow-diagram" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="taxprofiler_tube.png" class="img-fluid figure-img" alt="Visual overview of the nf-core/taxprofiler workflow. nf-core/taxprofiler can take in FASTQ (short or long reads) or FASTA files (long reads), that will optionally go through sequencing quality control, read preprocessing, complexity filtering, host removal, and run merging before going into taxonomic classification and/or profiling with a user-selected range of tools and databases. Output from all classifiers and profilers are additionally standardised into a common taxon table format and some with visualisation of profiles."></p>
<figcaption class="figure-caption">Figure&nbsp;1: Visual overview of the nf-core/taxprofiler workflow. nf-core/taxprofiler can take in FASTQ (short or long reads) or FASTA files (long reads), that will optionally go through sequencing quality control, read preprocessing, complexity filtering, host removal, and run merging before going into taxonomic classification and/or profiling with a user-selected range of tools and databases. Output from all classifiers and profilers are additionally standardised into a common taxon table format and some with visualisation of profiles.</figcaption>
</figure>
</div>
<p>We note that while we are aware of distinctions between metagenomic taxonomic classifiers (or taxonomic binner ) and profilers <span class="citation" data-cites="Ye2019-fl Meyer2022-dg">(<a href="#ref-Ye2019-fl" role="doc-biblioref">Ye et al. 2019</a>; <a href="#ref-Meyer2022-dg" role="doc-biblioref">Meyer et al. 2022</a>)</span>, we will generally refer to classifiers and profilers as synonyms throughout this article, as previously described <span class="citation" data-cites="Beber2023-hk">(<a href="#ref-Beber2023-hk" role="doc-biblioref">Beber et al. 2023</a>)</span>, as many different tools of both types are used in the pipeline.</p>
<section id="input-and-execution" class="level3">
<h3 class="anchored" data-anchor-id="input-and-execution">Input and Execution</h3>
<p>The pipeline can be executed via typical Nextflow commands, or using the standard nf-core ‘launch’ GUI (<a href="https://nf-co.re/taxprofiler/launch">https://nf-co.re/taxprofiler/launch</a>), making the pipeline accessible for both computationally experienced as well as less experienced researchers. In addition to the general usage and parameter documentation of the pipeline (https://nf-co.re/taxprofiler). The GUI offers immediate assistance and guidance to users on what each parameter does, both in short- and long-form, with long-form parameter descriptions additionally describing which tool-specific parameter(s) are being modified for each pipeline parameter (<a href="#fig-launch-page">Figure&nbsp;2</a>). The GUI also includes controlled user input by providing strict drop-down lists and input validation prior execution of the pipeline to reduce the risk of typos and other mistakes (in contrast the command-line interface (CLI) that only includes validation at pipeline run-time).</p>
<div id="fig-launch-page" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="taxprofiler_launchpage.png" class="img-fluid figure-img" alt="Screenshot of the nf-core pipeline launch graphical user interface with nf-core/taxprofiler options displayed. The web browser-based interface provides guidance for how to configure each pipeline parameter by providing both short and long help descriptions to help guide users in which contexts to configure each parameter. Additional elements such as radio buttons, drop down menus, and background regular expressions check for validity of input. When pressing launch, a prepared configuration file and command is provided that can be copied and pasted by the user into the terminal"></p>
<figcaption class="figure-caption">Figure&nbsp;2: Screenshot of the nf-core pipeline launch graphical user interface with nf-core/taxprofiler options displayed. The web browser-based interface provides guidance for how to configure each pipeline parameter by providing both short and long help descriptions to help guide users in which contexts to configure each parameter. Additional elements such as radio buttons, drop down menus, and background regular expressions check for validity of input. When pressing launch, a prepared configuration file and command is provided that can be copied and pasted by the user into the terminal</figcaption>
</figure>
</div>
<p>An example nf-core command line execution of the pipeline can be seen in figure (<a href="#lst-example-cmd">Code Block&nbsp;1</a>), where two input files are supplied: one of paths of FASTQ files of metagenomic samples and necessary metadata for preprocessing (such as sample ID and sequencing platform), and then a second giving paths to the user-supplied databases with parameters the user wishes the databases to be aligned to with. Various parameters are available to optionally turn on different preprocessing steps, and provide additional configurations such as tool selection and value options. Even if a user supplies a tool’s database to the database input sheet, the user must still activate the execution of each classifier they wish to use with dedicated flags, with some additional per-classifier flags for saving particular optional output files.</p>
<div id="lst-example-cmd" class="listing">
<p>Listing&nbsp;1: Example nf-core/taxprofiler command for running short-read quality control, removal of host DNA and executing the k-mer based Kraken2 and marker gene alignment MetaPhlAn3 tools.</p>
<div class="sourceCode" id="lst-example-cmd" data-lst-cap="Example nf-core/taxprofiler command for running short-read quality control, removal of host DNA and executing the k-mer based Kraken2 and marker gene alignment MetaPhlAn3 tools."><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="lst-example-cmd-1"><a href="#lst-example-cmd-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> nextflow run nf-core/taxprofiler <span class="at">-r</span> 1.1.0 <span class="dt">\</span></span>
<span id="lst-example-cmd-2"><a href="#lst-example-cmd-2" aria-hidden="true" tabindex="-1"></a>-profile singularity,<span class="op">&lt;</span>institute<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="lst-example-cmd-3"><a href="#lst-example-cmd-3" aria-hidden="true" tabindex="-1"></a>--input <span class="op">&lt;</span>samplesheet.csv<span class="op">&gt;</span> --databases <span class="op">&lt;</span>database.csv<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="lst-example-cmd-4"><a href="#lst-example-cmd-4" aria-hidden="true" tabindex="-1"></a>--perform_shortread_qc  <span class="at">--shortread_qc_minlength</span> 20 <span class="at">--preprocessing_qc_tool</span> falco <span class="dt">\</span></span>
<span id="lst-example-cmd-5"><a href="#lst-example-cmd-5" aria-hidden="true" tabindex="-1"></a>--run_host_removal <span class="at">--hostremoval_reference</span> <span class="st">'host_genome.fasta'</span> <span class="dt">\</span></span>
<span id="lst-example-cmd-6"><a href="#lst-example-cmd-6" aria-hidden="true" tabindex="-1"></a>--run_kraken2 <span class="at">--kraken2_save_reads</span> <span class="dt">\</span></span>
<span id="lst-example-cmd-7"><a href="#lst-example-cmd-7" aria-hidden="true" tabindex="-1"></a>--run_metaphlan3 <span class="dt">\</span></span>
<span id="lst-example-cmd-8"><a href="#lst-example-cmd-8" aria-hidden="true" tabindex="-1"></a>--run_krona <span class="dt">\</span></span>
<span id="lst-example-cmd-9"><a href="#lst-example-cmd-9" aria-hidden="true" tabindex="-1"></a>--run_profile_standardisation <span class="dt">\</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All nf-core pipelines are strictly versioned (specified with the Nextflow <code>-r</code> flag), and to ensure reproducibility each version of the pipeline has a fixed set of software used for each step of the pipeline. The fixed set of software are controlled through the use of the conda package manager and containers (e.g., docker, singularity) from the stable Bioconda <span class="citation" data-cites="Gruning2018-vr">(<a href="#ref-Gruning2018-vr" role="doc-biblioref">Grüning et al. 2018</a>)</span> and BioContainers <span class="citation" data-cites="Da_Veiga_Leprevost2017-gl">(<a href="#ref-Da_Veiga_Leprevost2017-gl" role="doc-biblioref">Veiga Leprevost et al. 2017</a>)</span> repositories. This, coupled with the intrinsic Nextflow ability to execute on most infrastructure whether that is a local laptop (resource requirements permitting), traditional HPC, as well across common cloud providers also makes nf-core/taxprofiler a very portable pipeline that can be used across many contexts.</p>
</section>
<section id="preprocessing" class="level3">
<h3 class="anchored" data-anchor-id="preprocessing">Preprocessing</h3>
<p>Preprocessing steps in nf-core/taxprofiler are aimed at removing laboratory and sequencing artefacts that may influence taxonomically profiling, either for computing resource consumption and/or false-positive or false-negative classification reasons. First sequencing quality control with FastQC <span class="citation" data-cites="Andrews2010-pd">(<a href="#ref-Andrews2010-pd" role="doc-biblioref">Andrews 2010</a>)</span> or Falco <span class="citation" data-cites="De_Sena_Brandine2021-pi">(<a href="#ref-De_Sena_Brandine2021-pi" role="doc-biblioref">Sena Brandine and Smith 2021</a>)</span> is carried out. Falco was included in for reduced memory requirements in particular for long reads sequencing. Library adapter sequences can reduce accuracy during alignment and are sometimes contaminants inside reference genomes [CITE] CARP BLOG OR OTHERS], and paired-end merging may provide longer sequences that will allow more specific classification when paired-end alignment is not supported by a given classifier. For these tasks nf-core/taxprofiler can apply either fastp <span class="citation" data-cites="Chen2018-vg">(<a href="#ref-Chen2018-vg" role="doc-biblioref">Chen et al. 2018</a>)</span> or AdapterRemoval2 <span class="citation" data-cites="Schubert2016-qv">(<a href="#ref-Schubert2016-qv" role="doc-biblioref">Schubert, Lindgreen, and Orlando 2016</a>)</span> for short-reads, and Porechop <span class="citation" data-cites="Wick2017-cn">(<a href="#ref-Wick2017-cn" role="doc-biblioref">Wick et al. 2017</a>)</span> for Oxford Nanopore long-read data. For both short- and long-reads FastQC or Falco is run again to allow assessment on the performance of the adapter removal and/or pair-merging step.</p>
<p>Low complexity sequences, e.g.&nbsp;sequences containing long stretches mono- or di-nucleotide repeats provide little specific genetic information that contribute to taxonomic identification, as they can align to many different reference genomes [CITE]. Including such reads during profiling can increase run-time and memory usage for little gain, as during lowest-common-ancestor (LCA) classification steps they will be assigned to high-level taxonomic levels such as Kingdom. nf-core/taxprofiler offers removal of these reads through complexity filtering algorithms as provided by fastp, BBDuk <span class="citation" data-cites="Bushnell2022-pf">(<a href="#ref-Bushnell2022-pf" role="doc-biblioref">Bushnell 2022</a>)</span>, or PRINSEQ++ <span class="citation" data-cites="Cantu2019-lh">(<a href="#ref-Cantu2019-lh" role="doc-biblioref">Cantu, Sadural, and Edwards 2019</a>)</span>. Long read sequences often do not have such reads as lengths are sufficient enough to capture greater sequence diversity - but it is sometimes desired to only classify reads longer than a certain length - as these provide sufficient taxonomic information. Therefore nf-core/taxprofiler can remove contigs shorter than a user-defined length using Filtlong.</p>
<p>Another common preprocessing step in metagenomic studies is to remove possible host DNA. This can help speed up run-time, particularly for microbiome studies, where detection of only microbes are of interest. Furthermore, host-contamination of reference genomes on public databases is common [CITE], therefore removal of such sequences can also decrease the risk of false positive taxonomic identifications. If users wish to remove multiple hosts or other sequences, they can simply combine all reference genomes into a single FASTA reference file. Short-read host removal can be carried out with Bowtie2 <span class="citation" data-cites="Langmead2012-ik Langmead2019-ej">(<a href="#ref-Langmead2012-ik" role="doc-biblioref">Langmead and Salzberg 2012</a>; <a href="#ref-Langmead2019-ej" role="doc-biblioref">Langmead et al. 2019</a>)</span> and minimap2 <span class="citation" data-cites="Li2018-qq">(<a href="#ref-Li2018-qq" role="doc-biblioref">Li 2018</a>)</span>for long-reads, both in combination with SAMtools <span class="citation" data-cites="Li2009-wy Danecek2021-gj">(<a href="#ref-Li2009-wy" role="doc-biblioref">Li et al. 2009</a>; <a href="#ref-Danecek2021-gj" role="doc-biblioref">Danecek et al. 2021</a>)</span>.</p>
<p>Finally, nf-core/taxprofiler can optionally perform run merging where libraries have been sequenced over multiple lanes to generate one profile per sample or library. The final set of reads used for profiling can be optionally saved for downstream re-use</p>
<p>Across all steps, relevant statistics and log files are generated and used both for the final pipeline run report as well as saved into the results directory of the pipeline run for deeper inspection where necessary.</p>
</section>
<section id="profiling" class="level3">
<h3 class="anchored" data-anchor-id="profiling">Profiling</h3>
<p>There are many types of metagenomic profile techniques from profiling against whole-genome references with alignment or k-mer based approaches, or methods involving alignment to species-specific marker-gene families <span class="citation" data-cites="Quince2017-bl">Ye et al. (<a href="#ref-Ye2019-fl" role="doc-biblioref">2019</a>)</span>. nf-core/taxprofiler aims to support and include all established classification or profiling tools as requested by the community. Which tools are used in a pipeline run is up to the user, with a tool being executed both a corresponding database and <code>--run_&lt;tool&gt;</code> flag is provided. Specific classification/profiling settings for each tool and database are specified in the database CSV input sheet, however for some tools pipeline level command-line flags are available.</p>
<p>As of version 1.1.0, the following classifiers and profilers are available: Kraken2 <span class="citation" data-cites="Wood2019-mf">(<a href="#ref-Wood2019-mf" role="doc-biblioref">Wood, Lu, and Langmead 2019</a>)</span>, Bracken <span class="citation" data-cites="Lu2017-vc">(<a href="#ref-Lu2017-vc" role="doc-biblioref">Lu et al. 2017</a>)</span>, KrakenUniq <span class="citation" data-cites="Breitwieser2018-xg">(<a href="#ref-Breitwieser2018-xg" role="doc-biblioref">F. P. Breitwieser, Baker, and Salzberg 2018</a>)</span>, Centrifuge <span class="citation" data-cites="Kim2016-qc">(<a href="#ref-Kim2016-qc" role="doc-biblioref">Kim et al. 2016</a>)</span>, MALT <span class="citation" data-cites="Vagene2018-px">(<a href="#ref-Vagene2018-px" role="doc-biblioref">Vågene et al. 2018</a>)</span>, DIAMOND <span class="citation" data-cites="Buchfink2021-ks">(<a href="#ref-Buchfink2021-ks" role="doc-biblioref">Buchfink, Reuter, and Drost 2021</a>)</span>, Kaiju <span class="citation" data-cites="Menzel2016-xy">(<a href="#ref-Menzel2016-xy" role="doc-biblioref">Menzel, Ng, and Krogh 2016</a>)</span>, MetaPhlAn3 and 4 <span class="citation" data-cites="Beghini2021-cf Blanco-Miguez2023-cq">(<a href="#ref-Beghini2021-cf" role="doc-biblioref">Beghini et al. 2021</a>; <a href="#ref-Blanco-Miguez2023-cq" role="doc-biblioref">Blanco-Míguez et al. 2023</a>)</span> mOTUs <span class="citation" data-cites="Ruscheweyh2022-hn">(<a href="#ref-Ruscheweyh2022-hn" role="doc-biblioref">Ruscheweyh et al. 2022</a>)</span>, ganon <span class="citation" data-cites="Piro2020-es">(<a href="#ref-Piro2020-es" role="doc-biblioref">Piro et al. 2020</a>)</span>, KMCP <span class="citation" data-cites="Shen2023-bk">(<a href="#ref-Shen2023-bk" role="doc-biblioref">Shen et al. 2023</a>)</span>. <a href="#tbl-tool-summaries">Table&nbsp;1</a> summarises the category and reference database type for each tool.</p>
<!-- TODO CHECK!!!!! AND add fig id AND maybe add classifier vs profiler? -->
<div id="tbl-tool-summaries" class="anchored">
<table class="table">
<caption>Table&nbsp;1: List of nf-core/taxprofiler supported taxonomic profilers as of version 1.1 and their method and support input database types. Sequencing matching type refers to which ‘alphabet’ is primarily used for matching between a query (read) and a reference (genome/gene). Method refers to the algorithm type used for matching. Reference type refers to the typical sequence type used in database construction of the tool.</caption>
<colgroup>
<col style="width: 34%">
<col style="width: 24%">
<col style="width: 23%">
<col style="width: 17%">
</colgroup>
<thead>
<tr class="header">
<th>Sequence Matching Type</th>
<th>Method</th>
<th>Reference Type</th>
<th>Tool</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Nucleotide</td>
<td>k-mer based</td>
<td>whole-genome</td>
<td>Kraken2</td>
</tr>
<tr class="even">
<td>Nucleotide</td>
<td>k-mer based</td>
<td>whole-genome</td>
<td>Bracken</td>
</tr>
<tr class="odd">
<td>Nucleotide</td>
<td>k-mer based</td>
<td>whole-genome</td>
<td>KrakenUniq</td>
</tr>
<tr class="even">
<td>Nucleotide</td>
<td>k-mer based</td>
<td>whole-genome</td>
<td>ganon</td>
</tr>
<tr class="odd">
<td>Nucleotide</td>
<td>k-mer based</td>
<td>whole-genome</td>
<td>KMCP</td>
</tr>
<tr class="even">
<td>Amino Acid</td>
<td>k-mer based</td>
<td>whole-genome</td>
<td>Kaiju</td>
</tr>
<tr class="odd">
<td>Nucleotide</td>
<td>alignment based</td>
<td>whole-genome</td>
<td>Centrifuge</td>
</tr>
<tr class="even">
<td>Nucleotide/Amino Acid</td>
<td>alignment based</td>
<td>whole-genome</td>
<td>MALT</td>
</tr>
<tr class="odd">
<td>Amino Acid</td>
<td>alignment based</td>
<td>whole-genome</td>
<td>DIAMOND</td>
</tr>
<tr class="even">
<td>Nucleotide</td>
<td>alignment based</td>
<td>marker-gene</td>
<td>MetaPhlAn3</td>
</tr>
<tr class="odd">
<td>Nucleotide</td>
<td>alignment based</td>
<td>marker-gene</td>
<td>MetaPhlAn4</td>
</tr>
<tr class="even">
<td>Nucleotide</td>
<td>alignment based</td>
<td>marker-gene</td>
<td>mOTUS</td>
</tr>
</tbody>
</table>
</div>
<!-- EMPHASISE PARALLELISATION AND EFFICIENCY -->
<p>nf-core/taxprofiler by default produces the per-sample main taxonomic classification profile from a tool or a tool’s report generation tool, normally in the form of counts per reference sequencing, with additional statistics about the hits of a particular organism (estimated abundance, taxonomic level etc.). Users can also optionally request output of per-read classification output, and output such as classified and unclassified reads in FASTQ format, where supported.</p>
</section>
<section id="post-profiling" class="level3">
<h3 class="anchored" data-anchor-id="post-profiling">Post-profiling</h3>
<p>Metagenomic studies often must compare the profiles between many samples, and the results of multiple profiles are normally stored in ‘taxon tables’, i.e, counts per reference (rows), for each sample (columns). When available nf-core/taxprofiler can optionally produce the ‘native’ taxon table of each classification tool when multiple samples are run.</p>
<p>When researchers wish to compare between taxonomic classifiers or profilers, they must often deal with the heterogenous output formats that are produced by each tool, typically in the form of custom parser and merging scripts for each tool. To facilitate more user-friendly cross-comparison between tools, nf-core/taxprofiler utilises the TAXPASTA tool <span class="citation" data-cites="Beber2023-hk">(<a href="#ref-Beber2023-hk" role="doc-biblioref">Beber et al. 2023</a>)</span> to generate standardised profiles and generate multi-sample tables.</p>
<p>Summary statistics for the entire pipeline are visualised are displayed in a customisable MultiQC report <span class="citation" data-cites="Ewels2020-vi">(<a href="#ref-Ewels2020-vi" role="doc-biblioref">Ewels et al. 2020</a>)</span>, when supported, to support user quality control of data and pipeline runs. Krona plots <span class="citation" data-cites="Ondov2011-se">(<a href="#ref-Ondov2011-se" role="doc-biblioref">Ondov, Bergman, and Phillippy 2011</a>)</span> can also optionally be generated for supported tools to help provide further visualisation of taxonomic profiles. <!-- TODO mention dynamic methods reporrting--></p>
</section>
<section id="output" class="level3">
<h3 class="anchored" data-anchor-id="output">Output</h3>
<p>To summarise, the main default output from nf-core/taxprofiler are both classifier ‘native’ and standardised single- and multi-sample taxonomic profiles with counts per-taxon and an interactive MultiQC run report with all run statistics, in addition to the raw log files themselves where available. The MultiQC run report displays statistics and summary visualisations for all steps of the pipeline where possible, and also includes versions for all tools and dynamically-constructed recommended publication methods text and citations (depending on what options were specified in the pipeline run) that users can use in their own publications. Optional outputs can include other types of profiles (e.g.&nbsp;per read classification) and in other formats as produced by the tools themselves, as well as raw reads from preprocessing steps and output visualisations from Krona. Nextflow resource usage and trace reports are also by default produced for users to check pipeline performance.</p>
<!-- TODO

## Use Cases and Benchmarking

To demonstrate different uses of nf-core/taxprofiler, we ran the pipeline on datasets of XXXX  from XXX , on a XXXX .

Run metric can be seen in Figure XXXXXX .

-->
</section>
</section>
<section id="discussion" class="level2">
<h2 class="anchored" data-anchor-id="discussion">Discussion</h2>
<!-- TODO: functionality comparison between other pipelines -->
<p>nf-core/taxprofiler has been designed strictly for the processing of <em>metagenomic</em> data. Other types of taxonomic profiling data such as 16S amplicon sequencing are well established fields with a range of popular high-quality and best-practise pipelines (e.g. <span class="citation" data-cites="Blanco-Miguez2023-cq Schloss2009-vk">(<a href="#ref-Blanco-Miguez2023-cq" role="doc-biblioref">Blanco-Míguez et al. 2023</a>; <a href="#ref-Schloss2009-vk" role="doc-biblioref">Schloss et al. 2009</a>)</span>) and databases <span class="citation" data-cites="DeSantis2006-cm Yilmaz2014-js">(<a href="#ref-DeSantis2006-cm" role="doc-biblioref">DeSantis et al. 2006</a>; <a href="#ref-Yilmaz2014-js" role="doc-biblioref">Yilmaz et al. 2014</a>)</span>, whereas the number of robust and high quality pipelines for metagenomics remains much lower - thus the need for highly-multiplexed classification is more desirable for the newer metagenomics method. Despite this, tools such as METAXA2 <span class="citation" data-cites="Bengtsson-Palme2015-ar">(<a href="#ref-Bengtsson-Palme2015-ar" role="doc-biblioref">Bengtsson-Palme et al. 2015</a>)</span> that use shotgun sequencing reads to recover 16S sequences from metagenomic samples.</p>
<p>The pipeline has been developed in the nf-core community, providing strong and long term support for the continued community-based development and maintenance of the pipeline. In this framework, we will continue to add additional metagenomic classification and profiling tools as they become established and as requested by the metagenomics community. This also applies to sequencing platforms, nf-core/taxprofiler already supports Nanopore long-read data, however long-read PacBio data <span class="citation" data-cites="Portik2022-jp">(<a href="#ref-Portik2022-jp" role="doc-biblioref">Portik, Brown, and Pierce-Ward 2022</a>)</span> may be of interest, and thus we could add appropriate preprocessing steps for this type of data too.</p>
<p>A remaining major challenge for metagenomics researchers is the construction of databases for each of the tools. Currently nf-core/taxprofiler requires users to construct these manually to supply to the pipeline (which we feel is currently reasonable investment as such databases can be repeatedly re-used), however we are exploring including an additional workflow in the pipeline to allow, given a set of FASTA reference files, automated database construction for all classification tool pipelines.</p>
<p>Finally, once an overall taxonomic profile is generated, researchers often wish to validate hits through more sensitive and accurate methods such as with read-mapping alignment. The pipeline could be easily extended to have a validation step where, utilising Nextflow’s execution parallelism, sequencing libraries could be aligned back to the reference genomes of only those species with hits from the taxonomic classification. This could also have additional benefits for researchers in palaeogenomics, where alignment information can be used to authenticate ancient DNA within their samples.</p>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>nf-core/taxprofiler is a accessible and efficient pipeline for metagenomic taxonomic classification and profiling that can be executed on anyway from laptops to the cloud provided there is sufficient computing resources. Offering a wide range of taxonomic profilers, it provides flexibility for users not just on choice of profiling tool but also with databases and database settings, with any number being able to be supplied to the pipeline in a single run. We hope that through detailed documentation and a range of execution options, that nf-core/taxprofiler will make reproducible and high-throughput metagenomics more accessible for a wide range of disciplines.</p>
<!-- TODO make this last sentence better, it's a bit poop -->
</section>
<section id="data-availability" class="level2">
<h2 class="anchored" data-anchor-id="data-availability">Data Availability</h2>
<p>All data used in this publication</p>
</section>
<section id="code-availability" class="level2">
<h2 class="anchored" data-anchor-id="code-availability">Code Availability</h2>
<p>nf-core/taxprofiler source code is available on GitHub at <a href="https://github.com/nf-core/taxprofiler">https://github.com/nf-core/taxprofiler</a>, and each release is archived on Zenodo (DOI: <a href="https://doi.org/10.5281/zenodo.7728364">10.5281/zenodo.7728364</a>)</p>
<p>The version of the pipeline described in this paper is version (<!-- VERSION -->) (Zenodo archive DOI: <!-- DOI -->)</p>
</section>
<section id="supplementary-data" class="level2">
<h2 class="anchored" data-anchor-id="supplementary-data">Supplementary Data</h2>
</section>
<section id="acknowledgments" class="level2">
<h2 class="anchored" data-anchor-id="acknowledgments">Acknowledgments</h2>
<p>We thank Prof.&nbsp;Christina Warinner and the Microbiome Sciences group MPI-EVA for original discussions that lead to the pipeline. We are also grateful for the nf-core community for ongoing support in the development in the pipeline, in particular for the contributions by Jianhong Ou.</p>
</section>
<section id="funding" class="level2">
<h2 class="anchored" data-anchor-id="funding">Funding</h2>
<p>S.S. and L.A-L. was supported by Rapid establishment of comprehensive laboratory pandemic preparedness – RAPID-SEQ This material is based upon work supported by the U.S. Department of Agriculture, Agricultural Research Service, under agreement No.&nbsp;58-3022-0-001 (T.A.C II). M.B. and J.A.F.Y were supported by the Max Planck Society. J.A.F.Y was supported by the Werner Siemens-Stiftung (“Paleobiotechnology”, Awarded to Prof.&nbsp;Pierre Stallforth and Prof.&nbsp;Christina Warinner).</p>
</section>
<section id="references" class="level2">




</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Andrews2010-pd" class="csl-entry" role="listitem">
Andrews, Simon. 2010. <span>“<span>FastQC</span>: A Quality Control Tool for High Throughput Sequence Data.”</span> <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a>.
</div>
<div id="ref-Beber2023-hk" class="csl-entry" role="listitem">
Beber, Moritz E, Maxime Borry, Sofia Stamouli, and James A Fellows Yates. 2023. <span>“<span>TAXPASTA</span>: <span>TAXonomic</span> Profile Aggregation and <span>STAndardisation</span>.”</span> <em>Journal of Open Source Software</em> 8 (87): 5627. <a href="https://doi.org/10.21105/joss.05627">https://doi.org/10.21105/joss.05627</a>.
</div>
<div id="ref-Beghini2021-cf" class="csl-entry" role="listitem">
Beghini, Francesco, Lauren J McIver, Aitor Blanco-Míguez, Leonard Dubois, Francesco Asnicar, Sagun Maharjan, Ana Mailyan, et al. 2021. <span>“Integrating Taxonomic, Functional, and Strain-Level Profiling of Diverse Microbial Communities with <span class="nocase">bioBakery</span> 3.”</span> <em>eLife</em> 10 (May): e65088. <a href="https://doi.org/10.7554/eLife.65088">https://doi.org/10.7554/eLife.65088</a>.
</div>
<div id="ref-Bengtsson-Palme2015-ar" class="csl-entry" role="listitem">
Bengtsson-Palme, Johan, Martin Hartmann, Karl Martin Eriksson, Chandan Pal, Kaisa Thorell, Dan Göran Joakim Larsson, and Rolf Henrik Nilsson. 2015. <span>“<span>METAXA2</span>: Improved Identification and Taxonomic Classification of Small and Large Subunit <span class="nocase">rRNA</span> in Metagenomic Data.”</span> <em>Molecular Ecology Resources</em> 15 (6): 1403–14. <a href="https://doi.org/10.1111/1755-0998.12399">https://doi.org/10.1111/1755-0998.12399</a>.
</div>
<div id="ref-Blanco-Miguez2023-cq" class="csl-entry" role="listitem">
Blanco-Míguez, Aitor, Francesco Beghini, Fabio Cumbo, Lauren J McIver, Kelsey N Thompson, Moreno Zolfo, Paolo Manghi, et al. 2023. <span>“Extending and Improving Metagenomic Taxonomic Profiling with Uncharacterized Species Using <span>MetaPhlAn</span> 4.”</span> <em>Nature Biotechnology</em>, February, 1–12. <a href="https://doi.org/10.1038/s41587-023-01688-w">https://doi.org/10.1038/s41587-023-01688-w</a>.
</div>
<div id="ref-Breitwieser2018-xg" class="csl-entry" role="listitem">
Breitwieser, F P, D N Baker, and S L Salzberg. 2018. <span>“<span>KrakenUniq</span>: Confident and Fast Metagenomics Classification Using Unique k-Mer Counts.”</span> <em>Genome Biology</em> 19 (1): 198. <a href="https://doi.org/10.1186/s13059-018-1568-0">https://doi.org/10.1186/s13059-018-1568-0</a>.
</div>
<div id="ref-Breitwieser2019-ys" class="csl-entry" role="listitem">
Breitwieser, Florian P, Jennifer Lu, and Steven L Salzberg. 2019. <span>“A Review of Methods and Databases for Metagenomic Classification and Assembly.”</span> <em>Briefings in Bioinformatics</em> 20 (4): 1125–36. <a href="https://doi.org/10.1093/bib/bbx120">https://doi.org/10.1093/bib/bbx120</a>.
</div>
<div id="ref-Buchfink2021-ks" class="csl-entry" role="listitem">
Buchfink, Benjamin, Klaus Reuter, and Hajk-Georg Drost. 2021. <span>“Sensitive Protein Alignments at Tree-of-Life Scale Using <span>DIAMOND</span>.”</span> <em>Nature Methods</em> 18 (4): 366–68. <a href="https://doi.org/10.1038/s41592-021-01101-x">https://doi.org/10.1038/s41592-021-01101-x</a>.
</div>
<div id="ref-Bushnell2022-pf" class="csl-entry" role="listitem">
Bushnell, Brian. 2022. <span>“<span>BBMap</span>.”</span> <a href="https://sourceforge.net/projects/bbmap/" class="uri">https://sourceforge.net/projects/bbmap/</a>.
</div>
<div id="ref-Cantu2019-lh" class="csl-entry" role="listitem">
Cantu, Vito Adrian, Jeffrey Sadural, and Robert Edwards. 2019. <span>“<span>PRINSEQ++</span>, a Multi-Threaded Tool for Fast and Efficient Quality Control and Preprocessing of Sequencing Datasets.”</span> e27553v1. PeerJ Preprints; PeerJ Inc. <a href="https://doi.org/10.7287/peerj.preprints.27553v1">https://doi.org/10.7287/peerj.preprints.27553v1</a>.
</div>
<div id="ref-Chen2018-vg" class="csl-entry" role="listitem">
Chen, Shifu, Yanqing Zhou, Yaru Chen, and Jia Gu. 2018. <span>“Fastp: An Ultra-Fast All-in-One <span>FASTQ</span> Preprocessor.”</span> <em>Bioinformatics</em> 34 (17): i884–90. <a href="https://doi.org/10.1093/bioinformatics/bty560">https://doi.org/10.1093/bioinformatics/bty560</a>.
</div>
<div id="ref-Danecek2021-gj" class="csl-entry" role="listitem">
Danecek, Petr, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu Ohan, Martin O Pollard, Andrew Whitwham, et al. 2021. <span>“Twelve Years of <span>SAMtools</span> and <span>BCFtools</span>.”</span> <em>GigaScience</em> 10 (2). <a href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a>.
</div>
<div id="ref-DeSantis2006-cm" class="csl-entry" role="listitem">
DeSantis, T Z, P Hugenholtz, N Larsen, M Rojas, E L Brodie, K Keller, T Huber, D Dalevi, P Hu, and G L Andersen. 2006. <span>“Greengenes, a Chimera-Checked <span>16S</span> <span class="nocase">rRNA</span> Gene Database and Workbench Compatible with <span>ARB</span>.”</span> <em>Applied and Environmental Microbiology</em> 72 (7): 5069–72. <a href="https://doi.org/10.1128/AEM.03006-05">https://doi.org/10.1128/AEM.03006-05</a>.
</div>
<div id="ref-Di_Tommaso2017-xu" class="csl-entry" role="listitem">
Di Tommaso, Paolo, Maria Chatzou, Evan W Floden, Pablo Prieto Barja, Emilio Palumbo, and Cedric Notredame. 2017. <span>“Nextflow Enables Reproducible Computational Workflows.”</span> <em>Nature Biotechnology</em> 35 (4): 316–19. <a href="https://doi.org/10.1038/nbt.3820">https://doi.org/10.1038/nbt.3820</a>.
</div>
<div id="ref-Eloe-Fadrosh2016-gz" class="csl-entry" role="listitem">
Eloe-Fadrosh, Emiley A, Natalia N Ivanova, Tanja Woyke, and Nikos C Kyrpides. 2016. <span>“Metagenomics Uncovers Gaps in Amplicon-Based Detection of Microbial Diversity.”</span> <em>Nature Microbiology</em> 1 (4): 15032. <a href="https://doi.org/10.1038/nmicrobiol.2015.32">https://doi.org/10.1038/nmicrobiol.2015.32</a>.
</div>
<div id="ref-Ewels2020-vi" class="csl-entry" role="listitem">
Ewels, Philip A, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso, and Sven Nahnsen. 2020. <span>“The Nf-Core Framework for Community-Curated Bioinformatics Pipelines.”</span> <em>Nature Biotechnology</em> 38 (3): 276–78. <a href="https://doi.org/10.1038/s41587-020-0439-x">https://doi.org/10.1038/s41587-020-0439-x</a>.
</div>
<div id="ref-Govender2022-td" class="csl-entry" role="listitem">
Govender, Kumeren N, and David W Eyre. 2022. <span>“Benchmarking Taxonomic Classifiers with Illumina and Nanopore Sequence Data for Clinical Metagenomic Diagnostic Applications.”</span> <em>Microbial Genomics</em> 8 (10): 000886. <a href="https://doi.org/10.1099/mgen.0.000886">https://doi.org/10.1099/mgen.0.000886</a>.
</div>
<div id="ref-Gruning2018-vr" class="csl-entry" role="listitem">
Grüning, Björn, Ryan Dale, Andreas Sjödin, Brad A Chapman, Jillian Rowe, Christopher H Tomkins-Tinch, Renan Valieris, Johannes Köster, and Bioconda Team. 2018. <span>“Bioconda: Sustainable and Comprehensive Software Distribution for the Life Sciences.”</span> <em>Nature Methods</em> 15 (7): 475–76. <a href="https://doi.org/10.1038/s41592-018-0046-7">https://doi.org/10.1038/s41592-018-0046-7</a>.
</div>
<div id="ref-Hillmann2018-ud" class="csl-entry" role="listitem">
Hillmann, Benjamin, Gabriel A Al-Ghalith, Robin R Shields-Cutler, Qiyun Zhu, Daryl M Gohl, Kenneth B Beckman, Rob Knight, and Dan Knights. 2018. <span>“Evaluating the Information Content of Shallow Shotgun Metagenomics.”</span> <em>mSystems</em> 3 (6). <a href="https://doi.org/10.1128/mSystems.00069-18">https://doi.org/10.1128/mSystems.00069-18</a>.
</div>
<div id="ref-Kim2016-qc" class="csl-entry" role="listitem">
Kim, Daehwan, Li Song, Florian P Breitwieser, and Steven L Salzberg. 2016. <span>“Centrifuge: Rapid and Sensitive Classification of Metagenomic Sequences.”</span> <em>Genome Research</em> 26 (12): 1721–29. <a href="https://doi.org/10.1101/gr.210641.116">https://doi.org/10.1101/gr.210641.116</a>.
</div>
<div id="ref-Langmead2012-ik" class="csl-entry" role="listitem">
Langmead, Ben, and Steven L Salzberg. 2012. <span>“Fast Gapped-Read Alignment with Bowtie 2.”</span> <em>Nature Methods</em> 9 (4): 357–59. <a href="https://doi.org/10.1038/nmeth.1923">https://doi.org/10.1038/nmeth.1923</a>.
</div>
<div id="ref-Langmead2019-ej" class="csl-entry" role="listitem">
Langmead, Ben, Christopher Wilks, Valentin Antonescu, and Rone Charles. 2019. <span>“Scaling Read Aligners to Hundreds of Threads on General-Purpose Processors.”</span> <em>Bioinformatics</em> 35 (3): 421–32. <a href="https://doi.org/10.1093/bioinformatics/bty648">https://doi.org/10.1093/bioinformatics/bty648</a>.
</div>
<div id="ref-Li2018-qq" class="csl-entry" role="listitem">
Li, Heng. 2018. <span>“<span>Minimap2</span>: Pairwise Alignment for Nucleotide Sequences.”</span> <em>Bioinformatics</em> 34 (18): 3094–3100. <a href="https://doi.org/10.1093/bioinformatics/bty191">https://doi.org/10.1093/bioinformatics/bty191</a>.
</div>
<div id="ref-Li2009-wy" class="csl-entry" role="listitem">
Li, Heng, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils Homer, Gabor Marth, Goncalo Abecasis, Richard Durbin, and 1000 Genome Project Data Processing Subgroup. 2009. <span>“The Sequence Alignment/Map Format and <span>SAMtools</span>.”</span> <em>Bioinformatics</em> 25 (16): 2078–79. <a href="https://doi.org/10.1093/bioinformatics/btp352">https://doi.org/10.1093/bioinformatics/btp352</a>.
</div>
<div id="ref-Lu2017-vc" class="csl-entry" role="listitem">
Lu, Jennifer, Florian P Breitwieser, Peter Thielen, and Steven L Salzberg. 2017. <span>“Bracken: Estimating Species Abundance in Metagenomics Data.”</span> <em>PeerJ. Computer Science</em> 3 (e104): e104. <a href="https://doi.org/10.7717/peerj-cs.104">https://doi.org/10.7717/peerj-cs.104</a>.
</div>
<div id="ref-Lynch2015-mz" class="csl-entry" role="listitem">
Lynch, Michael D J, and Josh D Neufeld. 2015. <span>“Ecology and Exploration of the Rare Biosphere.”</span> <em>Nature Reviews. Microbiology</em> 13 (4): 217–29. <a href="https://doi.org/10.1038/nrmicro3400">https://doi.org/10.1038/nrmicro3400</a>.
</div>
<div id="ref-McIntyre2017-td" class="csl-entry" role="listitem">
McIntyre, Alexa B R, Rachid Ounit, Ebrahim Afshinnekoo, Robert J Prill, Elizabeth Hénaff, Noah Alexander, Samuel S Minot, et al. 2017. <span>“Comprehensive Benchmarking and Ensemble Approaches for Metagenomic Classifiers.”</span> <em>Genome Biology</em> 18 (1): 182. <a href="https://doi.org/10.1186/s13059-017-1299-7">https://doi.org/10.1186/s13059-017-1299-7</a>.
</div>
<div id="ref-Menzel2016-xy" class="csl-entry" role="listitem">
Menzel, Peter, Kim Lee Ng, and Anders Krogh. 2016. <span>“Fast and Sensitive Taxonomic Classification for Metagenomics with Kaiju.”</span> <em>Nature Communications</em> 7 (April): 11257. <a href="https://doi.org/10.1038/ncomms11257">https://doi.org/10.1038/ncomms11257</a>.
</div>
<div id="ref-Meyer2022-dg" class="csl-entry" role="listitem">
Meyer, Fernando, Adrian Fritz, Zhi-Luo Deng, David Koslicki, Till Robin Lesker, Alexey Gurevich, Gary Robertson, et al. 2022. <span>“Critical Assessment of Metagenome Interpretation: The Second Round of Challenges.”</span> <em>Nature Methods</em> 19 (4): 429–40. <a href="https://doi.org/10.1038/s41592-022-01431-4">https://doi.org/10.1038/s41592-022-01431-4</a>.
</div>
<div id="ref-Mitchell2019-kt" class="csl-entry" role="listitem">
Mitchell, Alex L, Alexandre Almeida, Martin Beracochea, Miguel Boland, Josephine Burgin, Guy Cochrane, Michael R Crusoe, et al. 2019. <span>“<span>MGnify</span>: The Microbiome Analysis Resource in 2020.”</span> <em>Nucleic Acids Research</em>, November. <a href="https://doi.org/10.1093/nar/gkz1035">https://doi.org/10.1093/nar/gkz1035</a>.
</div>
<div id="ref-Nasko2018-cl" class="csl-entry" role="listitem">
Nasko, Daniel J, Sergey Koren, Adam M Phillippy, and Todd J Treangen. 2018. <span>“<span>RefSeq</span> Database Growth Influences the Accuracy of k-Mer-Based Lowest Common Ancestor Species Identification.”</span> <em>Genome Biology</em> 19 (1): 165. <a href="https://doi.org/10.1186/s13059-018-1554-6">https://doi.org/10.1186/s13059-018-1554-6</a>.
</div>
<div id="ref-Nayfach2016-ej" class="csl-entry" role="listitem">
Nayfach, Stephen, and Katherine S Pollard. 2016. <span>“Toward Accurate and Quantitative Comparative Metagenomics.”</span> <em>Cell</em> 166 (5): 1103–16. <a href="https://doi.org/10.1016/j.cell.2016.08.007">https://doi.org/10.1016/j.cell.2016.08.007</a>.
</div>
<div id="ref-Ondov2011-se" class="csl-entry" role="listitem">
Ondov, Brian D, Nicholas H Bergman, and Adam M Phillippy. 2011. <span>“Interactive Metagenomic Visualization in a Web Browser.”</span> <em>BMC Bioinformatics</em> 12 (1): 385. <a href="https://doi.org/10.1186/1471-2105-12-385">https://doi.org/10.1186/1471-2105-12-385</a>.
</div>
<div id="ref-Piro2020-es" class="csl-entry" role="listitem">
Piro, Vitor C, Temesgen H Dadi, Enrico Seiler, Knut Reinert, and Bernhard Y Renard. 2020. <span>“Ganon: Precise Metagenomics Classification Against Large and up-to-Date Sets of Reference Sequences.”</span> <em>Bioinformatics (Oxford, England)</em> 36 (Suppl_1): i12–20. <a href="https://doi.org/10.1093/bioinformatics/btaa458">https://doi.org/10.1093/bioinformatics/btaa458</a>.
</div>
<div id="ref-Portik2022-jp" class="csl-entry" role="listitem">
Portik, Daniel M, C Titus Brown, and N Tessa Pierce-Ward. 2022. <span>“Evaluation of Taxonomic Classification and Profiling Methods for Long-Read Shotgun Metagenomic Sequencing Datasets.”</span> <em>BMC Bioinformatics</em> 23 (1): 541. <a href="https://doi.org/10.1186/s12859-022-05103-0">https://doi.org/10.1186/s12859-022-05103-0</a>.
</div>
<div id="ref-Quince2017-bl" class="csl-entry" role="listitem">
Quince, Christopher, Alan W Walker, Jared T Simpson, Nicholas J Loman, and Nicola Segata. 2017. <span>“Shotgun Metagenomics, from Sampling to Analysis.”</span> <em>Nature Biotechnology</em> 35 (9): 833–44. <a href="https://doi.org/10.1038/nbt.3935">https://doi.org/10.1038/nbt.3935</a>.
</div>
<div id="ref-Ruscheweyh2022-hn" class="csl-entry" role="listitem">
Ruscheweyh, Hans-Joachim, Alessio Milanese, Lucas Paoli, Nicolai Karcher, Quentin Clayssen, Marisa Isabell Keller, Jakob Wirbel, et al. 2022. <span>“Cultivation-Independent Genomes Greatly Expand Taxonomic-Profiling Capabilities of <span class="nocase">mOTUs</span> Across Various Environments.”</span> <em>Microbiome</em> 10 (1): 212. <a href="https://doi.org/10.1186/s40168-022-01410-z">https://doi.org/10.1186/s40168-022-01410-z</a>.
</div>
<div id="ref-Schloss2009-vk" class="csl-entry" role="listitem">
Schloss, Patrick D, Sarah L Westcott, Thomas Ryabin, Justine R Hall, Martin Hartmann, Emily B Hollister, Ryan A Lesniewski, et al. 2009. <span>“Introducing Mothur: Open-Source, Platform-Independent, Community-Supported Software for Describing and Comparing Microbial Communities.”</span> <em>Applied and Environmental Microbiology</em> 75 (23): 7537–41. <a href="https://doi.org/10.1128/AEM.01541-09">https://doi.org/10.1128/AEM.01541-09</a>.
</div>
<div id="ref-Schubert2016-qv" class="csl-entry" role="listitem">
Schubert, Mikkel, Stinus Lindgreen, and Ludovic Orlando. 2016. <span>“<span>AdapterRemoval</span> <span class="nocase">v2</span>: Rapid Adapter Trimming, Identification, and Read Merging.”</span> <em>BMC Research Notes</em> 9 (February): 88. <a href="https://doi.org/10.1186/s13104-016-1900-2">https://doi.org/10.1186/s13104-016-1900-2</a>.
</div>
<div id="ref-Sczyrba2017-ma" class="csl-entry" role="listitem">
Sczyrba, Alexander, Peter Hofmann, Peter Belmann, David Koslicki, Stefan Janssen, Johannes Dröge, Ivan Gregor, et al. 2017. <span>“Critical Assessment of Metagenome Interpretation-a Benchmark of Metagenomics Software.”</span> <em>Nature Methods</em> 14 (11): 1063–71. <a href="https://doi.org/10.1038/nmeth.4458">https://doi.org/10.1038/nmeth.4458</a>.
</div>
<div id="ref-De_Sena_Brandine2021-pi" class="csl-entry" role="listitem">
Sena Brandine, Guilherme de, and Andrew D Smith. 2021. <span>“Falco: High-Speed <span>FastQC</span> Emulation for Quality Control of Sequencing Data.”</span> <em>F1000Research</em> 8 (1874): 1874. <a href="https://doi.org/10.12688/f1000research.21142.2">https://doi.org/10.12688/f1000research.21142.2</a>.
</div>
<div id="ref-Sharpton2014-wy" class="csl-entry" role="listitem">
Sharpton, Thomas J. 2014. <span>“An Introduction to the Analysis of Shotgun Metagenomic Data.”</span> <em>Frontiers in Plant Science</em> 5 (June): 209. <a href="https://doi.org/10.3389/fpls.2014.00209">https://doi.org/10.3389/fpls.2014.00209</a>.
</div>
<div id="ref-Shen2023-bk" class="csl-entry" role="listitem">
Shen, Wei, Hongyan Xiang, Tianquan Huang, Hui Tang, Mingli Peng, Dachuan Cai, Peng Hu, and Hong Ren. 2023. <span>“<span>KMCP</span>: Accurate Metagenomic Profiling of Both Prokaryotic and Viral Populations by Pseudo-Mapping.”</span> <em>Bioinformatics</em> 39 (1): btac845. <a href="https://doi.org/10.1093/bioinformatics/btac845">https://doi.org/10.1093/bioinformatics/btac845</a>.
</div>
<div id="ref-Sun2021-tq" class="csl-entry" role="listitem">
Sun, Zheng, Shi Huang, Meng Zhang, Qiyun Zhu, Niina Haiminen, Anna Paola Carrieri, Yoshiki Vázquez-Baeza, et al. 2021. <span>“Challenges in Benchmarking Metagenomic Profilers.”</span> <em>Nature Methods</em> 18 (6): 618–26. <a href="https://doi.org/10.1038/s41592-021-01141-3">https://doi.org/10.1038/s41592-021-01141-3</a>.
</div>
<div id="ref-Vagene2018-px" class="csl-entry" role="listitem">
Vågene, Åshild J, Alexander Herbig, Michael G Campana, Nelly M Robles García, Christina Warinner, Susanna Sabin, Maria A Spyrou, et al. 2018. <span>“Salmonella Enterica Genomes from Victims of a Major Sixteenth-Century Epidemic in Mexico.”</span> <em>Nature Ecology &amp; Evolution</em> 2 (3): 520–28. <a href="https://doi.org/10.1038/s41559-017-0446-6">https://doi.org/10.1038/s41559-017-0446-6</a>.
</div>
<div id="ref-Da_Veiga_Leprevost2017-gl" class="csl-entry" role="listitem">
Veiga Leprevost, Felipe da, Björn A Grüning, Saulo Alves Aflitos, Hannes L Röst, Julian Uszkoreit, Harald Barsnes, Marc Vaudel, et al. 2017. <span>“<span>BioContainers</span>: An Open-Source and Community-Driven Framework for Software Standardization.”</span> <em>Bioinformatics (Oxford, England)</em> 33 (16): 2580–82. <a href="https://doi.org/10.1093/bioinformatics/btx192">https://doi.org/10.1093/bioinformatics/btx192</a>.
</div>
<div id="ref-Wick2017-cn" class="csl-entry" role="listitem">
Wick, Ryan R, Louise M Judd, Claire L Gorrie, and Kathryn E Holt. 2017. <span>“Completing Bacterial Genome Assemblies with Multiplex <span>MinION</span> Sequencing.”</span> <em>Microbial Genomics</em> 3 (10): e000132. <a href="https://doi.org/10.1099/mgen.0.000132">https://doi.org/10.1099/mgen.0.000132</a>.
</div>
<div id="ref-Wood2019-mf" class="csl-entry" role="listitem">
Wood, Derrick E, Jennifer Lu, and Ben Langmead. 2019. <span>“Improved Metagenomic Analysis with Kraken 2.”</span> <em>Genome Biology</em> 20 (1): 257. <a href="https://doi.org/10.1186/s13059-019-1891-0">https://doi.org/10.1186/s13059-019-1891-0</a>.
</div>
<div id="ref-Wright2023-yr" class="csl-entry" role="listitem">
Wright, Robyn J, Andrè M Comeau, and Morgan G I Langille. 2023. <span>“From Defaults to Databases: Parameter and Database Choice Dramatically Impact the Performance of Metagenomic Taxonomic Classification Tools.”</span> <em>Microbial Genomics</em> 9 (3). <a href="https://doi.org/10.1099/mgen.0.000949">https://doi.org/10.1099/mgen.0.000949</a>.
</div>
<div id="ref-Ye2019-fl" class="csl-entry" role="listitem">
Ye, Simon H, Katherine J Siddle, Daniel J Park, and Pardis C Sabeti. 2019. <span>“Benchmarking Metagenomics Tools for Taxonomic Classification.”</span> <em>Cell</em> 178 (4): 779–94. <a href="https://doi.org/10.1016/j.cell.2019.07.010">https://doi.org/10.1016/j.cell.2019.07.010</a>.
</div>
<div id="ref-Yilmaz2014-js" class="csl-entry" role="listitem">
Yilmaz, Pelin, Laura Wegener Parfrey, Pablo Yarza, Jan Gerken, Elmar Pruesse, Christian Quast, Timmy Schweer, Jörg Peplies, Wolfgang Ludwig, and Frank Oliver Glöckner. 2014. <span>“The <span>SILVA</span> and <span>‘All-Species Living Tree Project (<span>LTP</span>)’</span> Taxonomic Frameworks.”</span> <em>Nucleic Acids Research</em> 42 (Database issue): D643–8. <a href="https://doi.org/10.1093/nar/gkt1209">https://doi.org/10.1093/nar/gkt1209</a>.
</div>
</div></section><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Reuse</h2><div id="quarto-reuse" class="quarto-appendix-contents"><div><a rel="license" href="https://creativecommons.org/licenses/by/4.0/">https://creativecommons.org/licenses/by/4.0/</a></div></div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>